### ISSUE TITLE
Accessing the open mHealth application inside containers from the Outside internet

### ISSUE THREAD

**Issue by manihatv** [openmhealth/shimmer#55]
Hi,

I am building an Android mobile application that makes use of the open mHealth's normalised health data collected from the third party APIs. For the authorization process, the application makes a call to the open mHealth's shimmer server. However, when authorizing a user from the android device, I came to understand that there are some issues in accessing the open mHealth application from the outside internet. 

I would like to know the correct way of accessing the open mHealth application(residing in the docker containers) from the outside internet. I am posting this query here so that I can get the help of the experts who have already tried this approach of authorizing users in omh server through Android application.

**Comment by emersonf** [comment]
Hi @manihatv,

Shimmer shouldn't be publicly available since it's not secured in any way (it's just an OAuth client, not a provider). The redirect should therefore hit your application, and your application should check/sanitize the redirect and forward it to Shimmer. So Shimmer is unreachable directly from the web.

The console's nginx file does this, but without any sanitisation.

Hope that helps.

**Comment by manihatv** [comment]
Hi @emersonf ,

Thanks for the response. I have tried the solution you suggested but I am still going wrong somewhere. Hence, I thought of explaining the situation in detail below, so that you can have a clear picture of the issue I am facing and suggest me with some suggestions accordingly. 

The primary issue I am facing here is that whenever I use an android device as emulator, I am facing issues in accessing the docker host. On the other hand, everything works perfectly fine when I run my application in an android emulator. 

I came to understand through a lot of discussion forums that the docker host cannot be accessed directly from the outside internet. A lot of people suggested the 'port forwarding' feature found under the adapter1 (within the Network tab of settings window) of the docker host. I then added a port forwarding row and this helped me to access the omh console with the ip address of my local host. That is, initially, I was accessing the console through the ip address of my docker host - "192.168.99.101:8083" and after implementing the port forwarding feature, I was able to access the console through the ipaddress of my local machine - " <localhost>:8083 ". I was then able to access the application from my android device as well.

Now, when I tried to run my android application in the android device, my application makes a call to the console -  "http://<localhost>:8083/authorize/{shim}?username={userId}" for the authorization purpose. As a response for the authorization, an 'authorizationurl' is sent in the JSON response and I am supposed to redirect the user to this 'authorizationurl'. This 'authorizationurl' has the ip address "192.168.99.101" which is the ip address of the docker host. As a result, when I try to redirect the user to this authorization URL, the control passes to the docker host where it fails to authorize, when accessing from the android device. 

On further analysing the issue, some of the discussion forums suggested using a bridged network connection between the parent host(the machine that I am using and where the virtual box is installed) and the docker host(the Linux machine installed in the Virtualbox and which contains the docker containers) as it allows the outside internet to access the docker host. So, I tried changing the network adapter2 of my docker host to 'bridged network'. The network adapter1 was kept unchanged as 'NAT'. After performing this step, I was not able to launch the docker container at all using the command " docker-machine start <docker-container> " . As a result, this suggestion also did not help me. 

And this issue has made me fall stagnant for a long time now. This issue may be a very small one for a lot of people out there. In that case, please do excuse me as I am a beginner.  

Thanks,
Viswa

**Comment by emersonf** [comment]
I assume you mean that the `authorizationUrl` contains `192.168.99.101` in the redirect URI parameter, correct?

If so, [this issue](https://github.com/openmhealth/shimmer/issues/38) contains a long explanation on working around that when using Docker.
If not, I'm not sure I follow, since the `authorizationUrl` itself should point to a data provider, like `https://runkeeper.com`.

**Comment by manihatv** [comment]
Hi @emersonf , 

Thank you for the suggestion. It helped me in going through some of the docker based files and helped me resolve the issue. However, I would like to know the impact of my workaround. 

The work around is that, I located the file "docker-compose.yml"  and changed the value of the parameter 'OPENMHEALTH_SHIM_SERVER_CALLBACKURLBASE'  from '192.168.99.101:8083' to '{my_local_machine_IPaddress}:8083'. I changed it as I was accessing the open mHealth console through my local machine ip address by enabling the port forwarding from the docker host.  

As of now, I am having a test environment where all the devices are connected to the same network. That is, my smartphone and the open mHealth console (which I am accessing through my local host IP address by port forwarding) are available in the same network. I would like to know the impact of my work around when all devices are connected in the same network. That is, will my workaround affect the omh functionality in any ways. 

Thanks.
